File: 백준/Gold/1062. 가르침/가르침.py
import sys

words_num, alphabet_num = map(int, sys.stdin.readline().strip().split())

if alphabet_num < 5:
    print(0)
    exit()

if alphabet_num == 26:
    print(words_num)
    exit()

words = []
for i in range(words_num):
    word = sys.stdin.readline().strip()
    words.append(set(word) - set('antic'))

legible_alpha = [0] * 26

for w in ('a', 'n', 't', 'i', 'c'):
    legible_alpha[ord(w) - ord('a')] = True

answer = 0


def dfs(idx: int, cnt: int) -> None:
    global answer

    if cnt == alphabet_num - 5:
        read_cnt = 0
        for word in words:
            flag = True

            for w in word:
                if not legible_alpha[ord(w) - ord('a')]:
                    flag = False
                    break

            if flag:
                read_cnt += 1

        answer = max(answer, read_cnt)
        return

    for i in range(idx, 26):
        if not legible_alpha[i]:
            legible_alpha[i] = True
            dfs(i, cnt + 1)
            legible_alpha[i] = False


dfs(0, 0)
print(answer)

File: 백준/Gold/1062. 가르침/가르침.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static int N, K;
    static String[] words;
    static boolean[] learned = new boolean[26];
    static int answer = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine().trim());

        N = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());

        if (K < 5) {
            System.out.println(0);
            return;
        }
        if (K == 26) {
            System.out.println(N);
            return;
        }

        words = new String[N];
        for (int i = 0; i < N; i++) {
            String s = br.readLine().trim();
            String filtered = s.replaceAll("[antic]", "");
            words[i] = filtered;
        }

        for (char c : new char[]{'a','n','t','i','c'}) {
            learned[c - 'a'] = true;
        }

        dfs(0, 0);
        System.out.println(answer);
    }

    static void dfs(int idx, int cnt) {
        if (cnt == K - 5) {
            int readCnt = 0;
            for (String word : words) {
                boolean canRead = true;
                for (int j = 0; j < word.length(); j++) {
                    char ch = word.charAt(j);
                    if (!learned[ch - 'a']) {
                        canRead = false;
                        break;
                    }
                }
                if (canRead) readCnt++;
            }
            answer = Math.max(answer, readCnt);
            return;
        }

        for (int i = idx; i < 26; i++) {
            if (!learned[i]) {
                learned[i] = true;
                dfs(i + 1, cnt + 1);
                learned[i] = false;
            }
        }
    }
}
